In [1]:
import pandas as pd
import numpy as np
import os
import glob
import import_ipynb
import folium
import functions_notebook


importing Jupyter notebook from functions_notebook.ipynb


### Site Metadata Review

In [2]:
## encoding error with site_metadata. This did not occur in R. Pulled encoding values from link below and can load successfully with cp437
## https://docs.python.org/3/library/codecs.html#standard-encodings

site_metadata = pd.read_csv('../data/site_metadata.csv', dtype='str', delimiter = ',', encoding='cp437')
site_metadata['filename'] = 'site_metadata.csv'
display(site_metadata.head())
display(site_metadata.tail())


source   site_no nhd_comid                        station_name site_type  \
0   USGS  05515500  13433691         KANKAKEE RIVER AT DAVIS, IN        ST   
1   USGS  05516665  13435155     YELLOW RIVER NEAR OAK GROVE, IN        ST   
2   USGS  05517000  13434009            YELLOW RIVER AT KNOX, IN        ST   
3   USGS  05517010  13435141         YELLOW RIVER NEAR BREMS, IN        ST   
4   USGS  05517500  13434643  KANKAKEE RIVER AT DUNNS BRIDGE, IN        ST   

      latitude    longitude       huc timezone station_nm           filename  
0  41.38963889  -86.7061667  07120001      EST        NaN  site_metadata.csv  
1  41.27138889  -86.4666667  07120001      EST        NaN  site_metadata.csv  
2   41.3028196  -86.6205695  07120001      EST        NaN  site_metadata.csv  
3  41.30666667     -86.7375  07120001      EST        NaN  site_metadata.csv  
4   41.2200392   -86.968356  07120001      EST        NaN  site_metadata.csv

source                                  site_no nhd_comid station_name  \
278  USGS_OFR  Fox River at Montgomery, IL (left bank)       NaN          NaN   
279  USGS_OFR     Fox River at Montgomery, IL (bridge)       NaN          NaN   
280  USGS_OFR    Fox River at Montgomery, IL (culvert)       NaN          NaN   
281  USGS_OFR                             private lake       NaN          NaN   
282  USGS_OFR                              Lake Storey       NaN          NaN   

    site_type latitude longitude  huc timezone  \
278        ST      NaN       NaN  NaN      NaN   
279        ST      NaN       NaN  NaN      NaN   
280        ST      NaN       NaN  NaN      NaN   
281        LK      NaN       NaN  NaN      NaN   
282        LK      NaN       NaN  NaN      NaN   

                                  station_nm           filename  
278  Fox River at Montgomery, IL (left bank)  site_metadata.csv  
279     Fox River at Montgomery, IL (bridge)  site_metadata.csv  
280    Fox River at Montgomery, IL (culvert)  site_metadata.csv  
281                             private lake  site_metadata.csv  
282                              Lake Storey  site_metadata.csv

### General Scan

In [3]:
## General Scan

print(site_metadata.shape)

functions_notebook.print_df_specs(site_metadata)

(283, 11)

 site_metadata.csv
<bound method DataFrame.info of        source                                  site_no nhd_comid  \
0        USGS                                 05515500  13433691   
1        USGS                                 05516665  13435155   
2        USGS                                 05517000  13434009   
3        USGS                                 05517010  13435141   
4        USGS                                 05517500  13434643   
..        ...                                      ...       ...   
278  USGS_OFR  Fox River at Montgomery, IL (left bank)       NaN   
279  USGS_OFR     Fox River at Montgomery, IL (bridge)       NaN   
280  USGS_OFR    Fox River at Montgomery, IL (culvert)       NaN   
281  USGS_OFR                             private lake       NaN   
282  USGS_OFR                              Lake Storey       NaN   

                           station_name site_type     latitude    longitude  \
0           KANKAKEE RIVER AT DAVIS, IN   

source   site_no nhd_comid                 station_name site_type  \
count     283       283       158                          169       283   
unique      5       283       147                          169         3   
top      USGS  05515500  13551315  KANKAKEE RIVER AT DAVIS, IN        ST   
freq      164         1         3                            1       182   

           latitude    longitude       huc timezone           station_nm  \
count           261          261       169      261                  114   
unique          261          258        18        2                  114   
top     41.38963889  -89.3855556  07120006      CST  Illinois River_D-30   
freq              1            2        30      242                    1   

                 filename  
count                 283  
unique                  1  
top     site_metadata.csv  
freq                  283


 Any duplicates: 

Empty DataFrame
Columns: [source, site_no, nhd_comid, station_name, site_type, latitude, longitude, huc, timezone, station_nm, filename]
Index: []




In [4]:
## Identifying NaN in the site_metadata 

print(site_metadata.info())

site_metadata.station_name[site_metadata.station_name.duplicated()]
len(site_metadata.station_name[site_metadata.station_name.isna()])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 283 entries, 0 to 282
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   source        283 non-null    object
 1   site_no       283 non-null    object
 2   nhd_comid     158 non-null    object
 3   station_name  169 non-null    object
 4   site_type     283 non-null    object
 5   latitude      261 non-null    object
 6   longitude     261 non-null    object
 7   huc           169 non-null    object
 8   timezone      261 non-null    object
 9   station_nm    114 non-null    object
 10  filename      283 non-null    object
dtypes: object(11)
memory usage: 24.4+ KB
None


114

In [5]:
site_metadata.source.value_counts()
site_metadata.timezone.unique()


array(['EST', 'CST', nan], dtype=object)

In [6]:
## Check unique duplicates. i.e. which are duplicated
site_metadata['station_name'][site_metadata['station_name'].duplicated()].unique()

array([nan], dtype=object)

In [7]:
print('number of unqiue station names: ', site_metadata['station_name'].nunique())
site_metadata['station_name'].unique()

number of unqiue station names:  169


array(['KANKAKEE RIVER AT DAVIS, IN', 'YELLOW RIVER NEAR OAK GROVE, IN',
       'YELLOW RIVER AT KNOX, IN', 'YELLOW RIVER NEAR BREMS, IN',
       'KANKAKEE RIVER AT DUNNS BRIDGE, IN',
       'KANKAKEE RIVER AT SHELBY, IN', 'KANKAKEE RIVER AT MOMENCE, IL',
       'IROQUOIS RIVER NEAR FORESMAN, IN',
       'IROQUOIS RIVER AT IROQUOIS, IL', 'SUGAR CREEK AT MILFORD, IL',
       'IROQUOIS RIVER NEAR CHEBANSE, IL',
       'KANKAKEE RIVER AT CUSTER PARK, IL',
       'KANKAKEE RIVER NEAR WILMINGTON, IL',
       'BRIGHTON CREEK AT BRIGHTON, WI',
       'NORTH MILL CREEK AT HICKORY CORNERS, IL',
       'HASTINGS CREEK NEAR LINDENHURST, IL',
       'NORTH MILL CREEK NEAR MILBURN, IL',
       'SALT CREEK AT WESTERN SPRINGS, IL',
       'DES PLAINES RIVER AT RIVERSIDE, IL',
       'NB CHICAGO RIVER AT GRAND AVENUE AT CHICAGO, IL',
       'CHICAGO RIVER AT COLUMBUS DRIVE AT CHICAGO, IL',
       'CHICAGO S & S CANAL AT WESTERN AVE AT CHICAGO, IL',
       'LITTLE CALUMET RIVER NEAR HAMMOND, IN', 'HART

- All sites are unique except NA


In [8]:
site_metadata.source.value_counts()

USGS        164
IEPA         92
USGS_OFR     22
IN033         4
USACE         1
Name: source, dtype: int64

In [9]:
## Check unique duplicates. i.e. which are duplicated
site_metadata[(site_metadata['source'] == 'IN033') & (site_metadata['station_name'].isna())]

site_metadata[(site_metadata['source'] == 'IEPA')].info()

## Has all lat lon
site_metadata[(site_metadata['source'] == 'IN033')]




<class 'pandas.core.frame.DataFrame'>
Int64Index: 92 entries, 169 to 260
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   source        92 non-null     object
 1   site_no       92 non-null     object
 2   nhd_comid     0 non-null      object
 3   station_name  0 non-null      object
 4   site_type     92 non-null     object
 5   latitude      92 non-null     object
 6   longitude     92 non-null     object
 7   huc           0 non-null      object
 8   timezone      92 non-null     object
 9   station_nm    92 non-null     object
 10  filename      92 non-null     object
dtypes: object(11)
memory usage: 8.6+ KB


source          site_no nhd_comid  \
149  IN033  413632087222001   3398182   
150  IN033  413651087284901  13463913   
151  IN033  413656087273701   3398162   
152  IN033  413705087281901  13464023   

                                          station_name site_type     latitude  \
149   GR CAL R E BRANCH W OF BRIDGE ST NR E CHICAGO IN        ST   41.6089722   
150      G CAL R W BR E OF INDPLS BLVD NR E CHICAGO IN        ST   41.6140361   
151  GR CAL R E BRANCH E OF KENNEDY AVE NR E CHICAG...        ST  41.61550556   
152     GRAND CAL R AT IN HARBOR CANAL NR E CHICAGO IN        ST  41.61796944   

       longitude       huc timezone station_nm           filename  
149   -87.372325  07120003      EST        NaN  site_metadata.csv  
150  -87.4801583  07120003      EST        NaN  site_metadata.csv  
151  -87.4602944  07120003      EST        NaN  site_metadata.csv  
152  -87.4719028  07120003      EST        NaN  site_metadata.csv

In [10]:
site_metadata[(site_metadata['source'] == 'USGS_OFR')]

source                                      site_no nhd_comid  \
261  USGS_OFR               Camp Walter Scott Lake (beach)       NaN   
262  USGS_OFR                            Wonder Lake (bay)       NaN   
263  USGS_OFR                          Wonder Lake (N bay)       NaN   
264  USGS_OFR                         Wonder Lake (SW bay)       NaN   
265  USGS_OFR       Fox River at Algonquin, IL (left bank)       NaN   
266  USGS_OFR      Fox River at Algonquin, IL (right bank)       NaN   
267  USGS_OFR                              Candlewick Lake       NaN   
268  USGS_OFR                  Candlewick Lake (boat ramp)       NaN   
269  USGS_OFR                        Candlewick Lake (bay)       NaN   
270  USGS_OFR                        Westlake (near beach)       NaN   
271  USGS_OFR                             Westlake (N bay)       NaN   
272  USGS_OFR                 Lake Le-Aqua-Na (near beach)       NaN   
273  USGS_OFR             Herrin-New Reservoir (boat ramp)       NaN   
274  USGS_OFR              PatriotÆs Park Lake (boat ramp)       NaN   
275  USGS_OFR                      Marion Lake (boat ramp)       NaN   
276  USGS_OFR  Vermilion River at Pontiac, IL (right bank)       NaN   
277  USGS_OFR   Vermilion River at Pontiac, IL (left bank)       NaN   
278  USGS_OFR      Fox River at Montgomery, IL (left bank)       NaN   
279  USGS_OFR         Fox River at Montgomery, IL (bridge)       NaN   
280  USGS_OFR        Fox River at Montgomery, IL (culvert)       NaN   
281  USGS_OFR                                 private lake       NaN   
282  USGS_OFR                                  Lake Storey       NaN   

    station_name site_type latitude longitude  huc timezone  \
261          NaN        LK      NaN       NaN  NaN      NaN   
262          NaN        LK      NaN       NaN  NaN      NaN   
263          NaN        LK      NaN       NaN  NaN      NaN   
264          NaN        LK      NaN       NaN  NaN      NaN   
265          NaN        ST      NaN       NaN  NaN      NaN   
266          NaN        ST      NaN       NaN  NaN      NaN   
267          NaN        LK      NaN       NaN  NaN      NaN   
268          NaN        LK      NaN       NaN  NaN      NaN   
269          NaN        LK      NaN       NaN  NaN      NaN   
270          NaN        LK      NaN       NaN  NaN      NaN   
271          NaN        LK      NaN       NaN  NaN      NaN   
272          NaN        LK      NaN       NaN  NaN      NaN   
273          NaN        LK      NaN       NaN  NaN      NaN   
274          NaN        LK      NaN       NaN  NaN      NaN   
275          NaN        LK      NaN       NaN  NaN      NaN   
276          NaN        ST      NaN       NaN  NaN      NaN   
277          NaN        ST      NaN       NaN  NaN      NaN   
278          NaN        ST      NaN       NaN  NaN      NaN   
279          NaN        ST      NaN       NaN  NaN      NaN   
280          NaN        ST      NaN       NaN  NaN      NaN   
281          NaN        LK      NaN       NaN  NaN      NaN   
282          NaN        LK      NaN       NaN  NaN      NaN   

                                      station_nm           filename  
261               Camp Walter Scott Lake (beach)  site_metadata.csv  
262                            Wonder Lake (bay)  site_metadata.csv  
263                          Wonder Lake (N bay)  site_metadata.csv  
264                         Wonder Lake (SW bay)  site_metadata.csv  
265       Fox River at Algonquin, IL (left bank)  site_metadata.csv  
266      Fox River at Algonquin, IL (right bank)  site_metadata.csv  
267                              Candlewick Lake  site_metadata.csv  
268                  Candlewick Lake (boat ramp)  site_metadata.csv  
269                        Candlewick Lake (bay)  site_metadata.csv  
270                        Westlake (near beach)  site_metadata.csv  
271                             Westlake (N bay)  site_metadata.csv  
272                 Lake Le-Aqua-Na (near beach)  site_metad

In [11]:
site_metadata[(site_metadata['source'] == 'IEPA')]
# all have lat lon but no site name
site_metadata[(site_metadata['source'] == 'IEPA')].info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 92 entries, 169 to 260
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   source        92 non-null     object
 1   site_no       92 non-null     object
 2   nhd_comid     0 non-null      object
 3   station_name  0 non-null      object
 4   site_type     92 non-null     object
 5   latitude      92 non-null     object
 6   longitude     92 non-null     object
 7   huc           0 non-null      object
 8   timezone      92 non-null     object
 9   station_nm    92 non-null     object
 10  filename      92 non-null     object
dtypes: object(11)
memory usage: 8.6+ KB


In [12]:
site_metadata.groupby('source').nunique()

site_no  nhd_comid  station_name  site_type  latitude  longitude  \
source                                                                       
IEPA           92          0             0          2        92         92   
IN033           4          4             4          1         4          4   
USACE           1          0             1          1         1          1   
USGS          164        144           164          3       164        161   
USGS_OFR       22          0             0          2         0          0   

          huc  timezone  station_nm  filename  
source                                         
IEPA        0         1          92         1  
IN033       1         1           0         1  
USACE       1         1           0         1  
USGS       18         2           0         1  
USGS_OFR    0         0          22         1

In [13]:
site_metadata[['source','station_name', 'latitude', 'longitude']][(site_metadata['source'] == 'USGS')].nunique(dropna= False)

source            1
station_name    164
latitude        164
longitude       161
dtype: int64

In [14]:
## Looking at non-usgs site info 

display(site_metadata[['source','station_name','latitude','longitude']]\
    [site_metadata['station_name'].isna()]\
    .sample(10)
       )
display(
site_metadata[['source','station_name','latitude','longitude']]\
    [site_metadata['source'] != 'USGS']\
    .station_name.unique()
) 

site_metadata[['source','station_name','latitude','longitude']]\
    [(site_metadata['source'] != 'USGS') & (~site_metadata['station_name'].isna())]

source station_name   latitude   longitude
267  USGS_OFR          NaN        NaN         NaN
227      IEPA          NaN   40.64938   -89.04137
257      IEPA          NaN    42.3572   -88.28788
279  USGS_OFR          NaN        NaN         NaN
197      IEPA          NaN  42.362472  -88.148871
233      IEPA          NaN  42.242947  -88.044172
201      IEPA          NaN  42.423261  -88.030081
269  USGS_OFR          NaN        NaN         NaN
234      IEPA          NaN   40.64877   -89.05465
243      IEPA          NaN   39.44309   -89.96278

array(['GR CAL R E BRANCH W OF BRIDGE ST NR E CHICAGO IN',
       'G CAL R W BR E OF INDPLS BLVD NR E CHICAGO IN',
       'GR CAL R E BRANCH E OF KENNEDY AVE NR E CHICAGO IN',
       'GRAND CAL R AT IN HARBOR CANAL NR E CHICAGO IN',
       'Starved Rock State Park, approximately .3 river miles upstream of Plum Island',
       nan], dtype=object)

source                                       station_name     latitude  \
149  IN033   GR CAL R E BRANCH W OF BRIDGE ST NR E CHICAGO IN   41.6089722   
150  IN033      G CAL R W BR E OF INDPLS BLVD NR E CHICAGO IN   41.6140361   
151  IN033  GR CAL R E BRANCH E OF KENNEDY AVE NR E CHICAG...  41.61550556   
152  IN033     GRAND CAL R AT IN HARBOR CANAL NR E CHICAGO IN  41.61796944   
168  USACE  Starved Rock State Park, approximately .3 rive...     41.32472   

       longitude  
149   -87.372325  
150  -87.4801583  
151  -87.4602944  
152  -87.4719028  
168    -88.98389

Station name unavailable for USGS_OFR and IEPA.

In [15]:
site_metadata[['source','station_name', 'latitude', 'longitude']][site_metadata['longitude'].isna()]


source station_name latitude longitude
261  USGS_OFR          NaN      NaN       NaN
262  USGS_OFR          NaN      NaN       NaN
263  USGS_OFR          NaN      NaN       NaN
264  USGS_OFR          NaN      NaN       NaN
265  USGS_OFR          NaN      NaN       NaN
266  USGS_OFR          NaN      NaN       NaN
267  USGS_OFR          NaN      NaN       NaN
268  USGS_OFR          NaN      NaN       NaN
269  USGS_OFR          NaN      NaN       NaN
270  USGS_OFR          NaN      NaN       NaN
271  USGS_OFR          NaN      NaN       NaN
272  USGS_OFR          NaN      NaN       NaN
273  USGS_OFR          NaN      NaN       NaN
274  USGS_OFR          NaN      NaN       NaN
275  USGS_OFR          NaN      NaN       NaN
276  USGS_OFR          NaN      NaN       NaN
277  USGS_OFR          NaN      NaN       NaN
278  USGS_OFR          NaN      NaN       NaN
279  USGS_OFR          NaN      NaN       NaN
280  USGS_OFR          NaN      NaN       NaN
281  USGS_OFR          NaN      NaN       NaN
282  USGS_OFR          NaN      NaN       NaN

The lat and lon are only unavailable for USGS_OFR site.


#### Explore site_no


In [16]:
## looking at how site_no differ based on source
site_metadata.groupby('source').site_no.unique()

source
IEPA        [D-30, D-05, G-08, DT-35, DT-22, DT-09, DT-06,...
IN033       [413632087222001, 413651087284901, 41365608727...
USACE           [Illinois River at Starved Rock Lock and Dam]
USGS        [05515500, 05516665, 05517000, 05517010, 05517...
USGS_OFR    [Camp Walter Scott Lake (beach), Wonder Lake (...
Name: site_no, dtype: object

#### Explore site geographical data

In [17]:
## Check the geographical data

locations = site_metadata[['latitude','longitude']].dropna()

locationlist = locations.values.tolist()


map = folium.Map(location=[41, -89], zoom_start=7)
for point in range(0, len(locationlist)):
    folium.Marker(locationlist[point], popup=f'Site displayed #{point}').add_to(map)

folium.Marker(locationlist[point], popup=f'Site displayed #{point}', icon=folium.Icon(color='red')).add_to(map)

display(map)

print(f'Only {len(locations)} can be displayed on the Map as 22 sites of USGS_OFR do not contain known lat long')


Only 261 can be displayed on the Map as 22 sites of USGS_OFR do not contain known lat long


In [18]:
## latitude min & max

functions_notebook.calc_metrics(site_metadata.latitude)

## longitude min & max

functions_notebook.calc_metrics(site_metadata.longitude)


df_col is of type object. Converted to float.
Col latitude values are: 
       min: 39.11008
       mean: 41.38494459888887
       max: 42.9269587
df_col is of type object. Converted to float.
Col longitude values are: 
       min: -90.8962451
       mean: -88.58199887417624
       max: -86.4666667


Latitude max, longitude min, and logitude max are equal to what is given in the habs metadata document. 
**Min latitude value is not equal to the metadata value given in metada document.** 

In [19]:
site_metadata[site_metadata.latitude.astype(float) == site_metadata.latitude.astype(float).min()]

source site_no nhd_comid station_name site_type  latitude  longitude  huc  \
177   IEPA    B-06       NaN          NaN        ST  39.11008  -87.65501  NaN   

    timezone         station_nm           filename  
177      CST  Wabash River_B-06  site_metadata.csv

In [20]:
## Checking huc column 

site_metadata.huc.value_counts(dropna=False)


NaN         114
07120006     30
07130001     29
07120004     22
07120003     16
07130009     13
07120001     12
07120007      7
07130007      6
07120005      6
07130003      6
07130012      5
07130002      4
07120002      4
07130006      3
07130011      3
07130005      1
07130010      1
07130008      1
Name: huc, dtype: int64

Value counts shows the 114 rows with no huc assocatiation - this 114 are all the non-USGS sites.

### Pcode group dataset review

In [21]:

## setting parameter_cd as str to avoid automatic read in as int/float
pcode_group_xwalk = pd.read_csv('../data/pcode_group_xwalk.csv',
                                dtype={'parameter_cd':'str'})

display(pcode_group_xwalk.head())

pcode_group_xwalk.shape


parameter_cd      generic_parm_name generic_parm_abbrev  min_allowed  \
0        49953  Phytoplankton biomass       phyto_biomass            0   
1        01325     Algal surface mats                mats            0   
2        32210          Chlorophyll a               chl_a            0   
3        70953          Chlorophyll a               chl_a            0   
4        70957          Chlorophyll a               chl_a            0   

   max_allowed units  
0          NaN  mg/l  
1          4.0  code  
2       1000.0  ug/l  
3       1000.0  ug/l  
4       1000.0  ug/l

(48, 6)

parameter_cd has trailing and leading 0 and should be interpreted as str when read in. updating read_csv function above. 

In [22]:
display(pcode_group_xwalk)

parameter_cd      generic_parm_name generic_parm_abbrev  min_allowed  \
0         49953  Phytoplankton biomass       phyto_biomass            0   
1         01325     Algal surface mats                mats            0   
2         32210          Chlorophyll a               chl_a            0   
3         70953          Chlorophyll a               chl_a            0   
4         70957          Chlorophyll a               chl_a            0   
5         32284           fChlorophyll                fchl            0   
6         32316           fChlorophyll                fchl            0   
7         32318           fChlorophyll                fchl            0   
8         32320           fChlorophyll                fchl            0   
9         62360             Pheophytin                pheo            0   
10        32319           fPhycocyanin                 fPC            0   
11        30208              Discharge                   Q            0   
12        30209              Discharge                   Q            0   
13        00061              Discharge                   Q            0   
14        00060   Continuous discharge        Q_continuous            0   
15        00065            Gage height         gage_height            0   
16        00072            Gage height         gage_height            0   
17        00055         Water velocity                 vel            0   
18        72149         Water velocity                 vel            0   
19        72190         Water velocity                 vel            0   
20        72196         Water velocity                 vel            0   
21        72254         Water velocity                 vel            0   
22        72255         Water velocity                 vel            0   
23        72322         Water velocity                 vel            0   
24        72323         Water velocity                 vel            0   
25        81380         Water velocity                 vel            0   
26        81904         Water velocity                 vel            0   
27        00618                Nitrate                 NO3            0   
28        00631   Nitrate plus nitrite             NO3_NO2            0   
29        99133   Nitrate plus nitrite             NO3_NO2            0   
30        00671         Orthophosphate                  OP            0   
31        51289         Orthophosphate                  OP            0   
32        32295                   fDOM                fDOM            0   
33        50624          UV absorbance              uv_254            0   
34        00095   Specific conductance                  SC            0   
35        00300       Dissolved oxygen                  DO            0   
36        00400                     pH                  pH            0   
37        00020        Air temperature            temp_air          -50   
38        00021        Air temperature            temp_air          -50   
39        00010      Water temperature          temp_water           -5   
40        00011      Water temperature          temp_water           -5   
41        00076              Turbidity                turb            0   
42        63675              Turbidity                turb            0   
43        63676              Turbidity                turb            0   
44        63680              Turbidity                turb            0   
45        80154               Sediment            sediment            0   
46        80180               Sediment            sediment            0   
47        00530               Sediment            sediment            0   

    max_allowed       units  
0           NaN        mg/l  
1           4.0        code  
2        1000.0        ug/l  
3        1000.0        ug/l  
4        1000.0        ug/l  
5         400.0        ug/l  
6         400.0        ug/l  
7         400.0        ug/l  
8         400.0        ug/l  
9        1000.0        ug/l

#### Pcode df Column exploration


In [23]:
## types and null count
pcode_group_xwalk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   parameter_cd         48 non-null     object 
 1   generic_parm_name    48 non-null     object 
 2   generic_parm_abbrev  48 non-null     object 
 3   min_allowed          48 non-null     int64  
 4   max_allowed          31 non-null     float64
 5   units                48 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 2.4+ KB


In [24]:
## count of uniques in each col

pcode_group_xwalk.nunique()

parameter_cd           48
generic_parm_name      22
generic_parm_abbrev    22
min_allowed             3
max_allowed            13
units                  14
dtype: int64

#### Looking at specific cols 

In [25]:
## Min and Max cols

## Count of unique min_allowed and max_allowed values
display(pcode_group_xwalk[['min_allowed','max_allowed']].value_counts(dropna = False))

# print(functions_notebook.calc_metrics(pcode_group_xwalk.min_allowed))
# print(functions_notebook.calc_metrics(pcode_group_xwalk.max_allowed))

functions_notebook.calc_metrics(pcode_group_xwalk.min_allowed)

min_allowed  max_allowed
 0           NaN            17
             400.0           4
             1000.0          4
             4000.0          4
             75000.0         3
-50          50.0            2
-5           50.0            2
 0           20.0            2
             100.0           2
             150.0           2
             1.0             1
             4.0             1
             14.0            1
             50.0            1
             300.0           1
             100000.0        1
dtype: int64

Col min_allowed values are: 
       min: -50
       mean: -2.2916666666666665
       max: 0


In [26]:
## Units Col

print(pcode_group_xwalk.units.unique())

print('\nUnique counts of given units:\n', pcode_group_xwalk.units.value_counts())

['mg/l' 'code' 'ug/l' 'm3/sec' 'm' 'm/sec' 'mg/l as N' 'mg/l as P'
 'ug/l QSE' 'units/cm' 'uS/cm @25C' 'std units' 'deg C' 'FNU']

Unique counts of given units:
 m/sec         10
ug/l           9
mg/l           5
m3/sec         4
deg C          4
FNU            4
mg/l as N      3
m              2
mg/l as P      2
code           1
ug/l QSE       1
units/cm       1
uS/cm @25C     1
std units      1
Name: units, dtype: int64
